gh-pages/content/adr/2021-01-21-go-override-syntax.md

```diff
@@ -0,0 +1,308 @@
+      return fmt.Sprintf("✨%s✨", g.DefaultGreeter.GetGreetee())
+    }
+
+    func (g *ExcitedGreeter) Greet() string {
+      return fmt.Sprintf("OMG %s!!!11!1!!!", g.DefaultGreeter.Greet())
+    }
+
+    ////////////////////
+    // main           //
+    ////////////////////
+    func main() {
+      boring := NewDefaultGreeter("Sarah Connors")
+      excited := NewExcitedGreeter("Sarah Connors")
+
+      fmt.Println(boring.Greet())  // #=> Hello, Sarah Connors!
+      fmt.Println(excited.Greet()) // #=> OMG Hello, ✨Sarah Connors✨!!!!11!1!!!
+    }
+    ```
+
+#### Kernel integration
+
+The *jsii runtime for go* must correctly register those overrides when creating
+the instance in the *jsii kernel*, and subsequently be able to invoke the
+correct *go* functions upon receiving a `callback` request.
+
+In order to allow subclasses to invoke their parent type's constructor without
+causing the `create` call to be prematurely sent over to the *jsii kernel*, a
+separate constructor function will need to be provided. This constructor will
+be similar to the "usual" one, except it will add the necessary `overrides`
+declarations to the `create` request parameters. It could also be responsible
+for setting the embedded interface value that enables dynamic dispatch.
+
+!!! example
+    ```go
+    // Pretend for a moment this is NOT an abstract class, and hence it can be
+    // constructed without being overridden... Then we would have a constructor
+    // somewhat like:
+    func NewGreeterBase() *GreeterBase {
+      o := &GreeterBase{}
+      jsii.Create(
+        "test.GreeterBase",
+        []interface{}, // None
+        []FQN{},       // None
+        []Override{},  // None
+        o,
+      )
+      return o
+    }
+
+    // NewGreeterBase_Override initializes the embedded value for `GreeterBase`
+    // and registers all overrides declared by the dynamic type of `i`. The
+    // resulting instance should be assigned to an anonymously embedded pointer
+    // of `GreeterBase`.
+    func NewGreeterBase_Override(i IGreeterBase) *GreeterBase {
+      o := &GreeterBase{}
+      o.IGreeterBase = i
+      jsii.Create(
+        "test.GreeterBase",
+        []interface{}, // None
+        []FQN{},       // None
+        determineOverrides(i),
+        i,
+      )
+      return o
+    }
+    ```
+
+As the **go** `reflect` library offers no facility to determine whether a method
+was promoted from an embedded type, or declared on the current type, the
+override detection cannot happen implicitly - the user is responsible to tag the
+embedded super-type with overrides.
+
+The proposed mechanism is to tag the embedded field with the `overrides` key,
+and a comma-delimited list of overridden methods as the value.
+
+??? example
+    ```go
+    type DefaultGreeter struct {
+      *GreeterBase `overrides:"GetGreetee"`
+      greetee      string
+    }
+
+    type ExcitedGreeter struct {
+      *DefaultGreeter `overrides:"GetGreetee,Greet"`
```

Comment:
Is it possible to put the tag on the method and not on the embedding?

Reply:
Methods cannot carry tags... Only struct fields.